File: Giki Chatbot Redit/chatbot.js
import fs from 'fs-extra';
import path from 'path';
import { fileURLToPath } from 'url';
import fetch from 'node-fetch';
import dotenv from 'dotenv';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

// Load environment variables
dotenv.config();

const INDEX_PATH = 'faiss_index';

// OpenRouter API Setup
const OPENROUTER_API_KEY = process.env.OPENROUTER_API_KEY;

// Reddit Setup (OAuth)
const REDDIT_CONFIG = {
    clientId: process.env.REDDIT_CLIENT_ID,
    clientSecret: process.env.REDDIT_CLIENT_SECRET,
    refreshToken: process.env.REDDIT_REFRESH_TOKEN,
    userAgent: process.env.REDDIT_USER_AGENT
};

// Document Processor
class GIKIDocumentProcessor {
    constructor(dataFolder = 'data') {
        this.dataFolder = dataFolder;
    }

    async extractTextFromPdf(filePath) {
        try {
            const pdfParse = await import('pdf-parse');
            const dataBuffer = await fs.readFile(filePath);
            const data = await pdfParse.default(dataBuffer);
            return data.text;
        } catch (error) {
            console.error(`Error processing PDF ${filePath}:`, error.message);
            return '';
        }
    }

    async extractTextFromDocx(filePath) {
        try {
            const mammoth = await import('mammoth');
            const result = await mammoth.extractRawText({ path: filePath });
            return result.value;
        } catch (error) {
            console.error(`Error processing DOCX ${filePath}:`, error.message);
            return '';
        }
    }

    async extractTextFromTxt(filePath) {
        try {
            return await fs.readFile(filePath, 'utf-8');
        } catch (error) {
            console.error(`Error processing TXT ${filePath}:`, error.message);
            return '';
        }
    }

    async extractTextFromJson(filePath) {
        try {
            const data = await fs.readJson(filePath);
            let text = '';
            
            if (Array.isArray(data)) {
                for (let i = 0; i < data.length; i++) {
                    const post = data[i];
                    if (typeof post === 'object' && post !== null) {
                        const title = post.title || '';
                        const selftext = post.selftext || '';
                        const author = post.author || '';
                        const postId = post.id || '';
                        
                        let postContent = `Title: ${title}\nContent: ${selftext}`;
                        
                        const comments = post.comments || [];
                        if (comments.length > 0) {
                            postContent += '\nComments:\n';
                            for (const comment of comments) {
                                if (typeof comment === 'object' && comment !== null) {
                                    const commentBody = comment.body || '';
                                    const commentAuthor = comment.author || '';
                                    if (commentBody) {
                                        postContent += `- ${commentAuthor}: ${commentBody}\n`;
                                    }
                                }
                            }
                        }
                        
                        text += `\n[Reddit Post ${i + 1} - ID: ${postId}]\n${postContent}\n`;
                    }
                }
            }
            
            return text;
        } catch (error) {
            console.error(`Error processing JSON ${filePath}:`, error.message);
            return '';
        }
    }

    async loadDocuments() {
        const documents = [];
        const dataPath = path.join(__dirname, this.dataFolder);

        if (!(await fs.pathExists(dataPath))) {
            return documents;
        }

        const supportedExtensions = ['.pdf', '.docx', '.txt', '.json'];
        const files = await fs.readdir(dataPath);

        for (const file of files) {
            const filePath = path.join(dataPath, file);
            const stats = await fs.stat(filePath);
            
            if (stats.isFile() && supportedExtensions.includes(path.extname(file).toLowerCase())) {
                let text = '';
                const ext = path.extname(file).toLowerCase();

                if (ext === '.pdf') {
                    text = await this.extractTextFromPdf(filePath);
                } else if (ext === '.docx') {
                    text = await this.extractTextFromDocx(filePath);
                } else if (ext === '.txt') {
                    text = await this.extractTextFromTxt(filePath);
                } else if (ext === '.json') {
                    text = await this.extractTextFromJson(filePath);
                }

                if (text.trim()) {
                    // Split text into chunks (simplified version)
                    const chunks = this.splitText(text);
                    for (let i = 0; i < chunks.length; i++) {
                        const chunk = chunks[i];
                        if (chunk.trim()) {
                            documents.push({
                                content: chunk,
                                metadata: {
                                    source: file,
                                    chunkId: i,
                                    fileType: ext
                                }
                            });
                        }
                    }
                }
            }
        }

        return documents;
    }

    splitText(text, chunkSize = 500, overlap = 50) {
        const chunks = [];
        const sentences = text.split(/[.!?]+/).filter(s => s.trim());
        let currentChunk = '';

        for (const sentence of sentences) {
            if ((currentChunk + sentence).length > chunkSize && currentChunk) {
                chunks.push(currentChunk.trim());
                currentChunk = sentence + '. ';
            } else {
                currentChunk += sentence + '. ';
            }
        }

        if (currentChunk.trim()) {
            chunks.push(currentChunk.trim());
        }

        return chunks;
    }
}

// Answer Quality Checker
class AnswerQualityChecker {
    constructor() {
        this.positiveIndicators = [
            'according to', 'based on', 'stated in', 'specified in', 'outlined in',
            'policy', 'rule', 'regulation', 'guideline', 'procedure',
            'prohibited', 'allowed', 'permitted', 'required', 'mandatory',
            'specific', 'detailed', 'comprehensive', 'complete', 'thorough'
        ];

        this.negativeIndicators = [
            'i don\'t have', 'i don\'t know', 'i cannot', 'not found', 'not specified',
            'no information', 'cannot find', 'unclear', 'not clear', 'sorry',
            'unable to', 'can\'t help', 'don\'t know', 'not available', 'missing'
        ];
    }

    calculateBasicScore(answer) {
        const answerLower = answer.toLowerCase().trim();
        const answerLength = answerLower.length;

        // Length score (0-1)
        const lengthScore = Math.min(answerLength / 200, 1.0);

        // Specificity score based on positive indicators
        const positiveCount = this.positiveIndicators.filter(indicator => 
            answerLower.includes(indicator)
        ).length;
        const specificityScore = Math.min(positiveCount / 3, 1.0);

        // Negativity score based on negative indicators
        const negativeCount = this.negativeIndicators.filter(indicator => 
            answerLower.includes(indicator)
        ).length;
        const negativityScore = Math.max(0, 1 - (negativeCount / 2));

        // Structure score (simplified)
        let structureScore = 0.0;
        if (answer.match(/[-•*]\s/)) structureScore += 0.3;
        if (answer.match(/\d+\.\s/)) structureScore += 0.2;
        if (answer.match(/[A-Z][a-z]+:/)) structureScore += 0.2;
        if (answer.match(/\*\*.*\*\*/)) structureScore += 0.1;
        if (answer.match(/\[.*\]/)) structureScore += 0.2;

        // Source citation score
        const citationScore = this.positiveIndicators.some(phrase => 
            answerLower.includes(phrase)
        ) ? 1.0 : 0.0;

        return {
            lengthScore,
            specificityScore,
            negativityScore,
            structureScore: Math.min(structureScore, 1.0),
            citationScore
        };
    }

    async getAIQualityAssessment(question, answer) {
        try {
            const prompt = `
            Assess the quality of this answer to the given question.
            
            Question: ${question}
            Answer: ${answer}
            
            Rate the answer on a scale of 1-10 for each criterion:
            1. Relevance: How well does it answer the question?
            2. Completeness: Does it provide sufficient information?
            3. Specificity: Is it specific and detailed?
            4. Helpfulness: Would this be useful to the user?
            
            Also determine if this answer is sufficient or if additional sources should be consulted.
            
            Respond in this exact format:
            Relevance: [1-10]
            Completeness: [1-10]
            Specificity: [1-10]
            Helpfulness: [1-10]
            Sufficient: [Yes/No]
            Reason: [Brief explanation]
            `;

            const response = await this.callOpenRouterAPI(prompt);
            
            // Parse the response
            const scores = {};
            let sufficient = false;
            let reason = '';

            const lines = response.split('\n');
            for (const line of lines) {
                if (line.includes(':')) {
                    const [key, value] = line.split(':', 2);
                    const keyTrim = key.trim();
                    const valueTrim = value.trim();

                    if (['Relevance', 'Completeness', 'Specificity', 'Helpfulness'].includes(keyTrim)) {
                        scores[keyTrim.toLowerCase()] = parseInt(valueTrim) || 5;
                    } else if (keyTrim === 'Sufficient') {
                        sufficient = valueTrim.toLowerCase() === 'yes';
                    } else if (keyTrim === 'Reason') {
                        reason = valueTrim;
                    }
                }
            }

            return {
                aiScores: scores,
                aiSufficient: sufficient,
                aiReason: reason
            };

        } catch (error) {
            console.error('AI assessment failed:', error.message);
            return {
                aiScores: { relevance: 5, completeness: 5, specificity: 5, helpfulness: 5 },
                aiSufficient: true,
                aiReason: 'AI assessment unavailable'
            };
        }
    }

    async assessAnswerQuality(question, answer) {
        // Basic scoring
        const basicScores = this.calculateBasicScore(answer);

        // AI assessment
        const aiAssessment = await this.getAIQualityAssessment(question, answer);

        // Calculate overall score
        const basicAvg = Object.values(basicScores).reduce((a, b) => a + b, 0) / Object.keys(basicScores).length;
        const aiAvg = Object.values(aiAssessment.aiScores).reduce((a, b) => a + b, 0) / Object.keys(aiAssessment.aiScores).length;

        // Weighted combination (70% AI, 30% basic)
        const overallScore = (aiAvg * 0.7) + (basicAvg * 0.3);

        // Determine if answer is sufficient
        const isSufficient = (
            overallScore >= 6.0 &&
            aiAssessment.aiSufficient &&
            basicScores.negativityScore >= 0.5
        );

        return {
            overallScore,
            isSufficient,
            basicScores,
            aiAssessment,
            recommendation: isSufficient ? 'sufficient' : 'needs_fallback'
        };
    }

    async callOpenRouterAPI(prompt) {
        const response = await fetch('https://openrouter.ai/api/v1/chat/completions', {
            method: 'POST',
            headers: {
                'Authorization': `Bearer ${OPENROUTER_API_KEY}`,
                'Content-Type': 'application/json',
                'HTTP-Referer': 'https://huggingface.co',
                'X-Title': 'GIKI-RAG-bot'
            },
            body: JSON.stringify({
                model: 'deepseek/deepseek-r1-0528-qwen3-8b:free',
                messages: [
                    {
                        role: 'system',
                        content: 'You are a helpful assistant for GIKI (Ghulam Ishaq Khan Institute of Engineering Sciences and Technology).'
                    },
                    {
                        role: 'user',
                        content: prompt
                    }
                ],
                temperature: 0.1
            })
        });

        if (!response.ok) {
            throw new Error(`OpenRouter API error: ${response.status}`);
        }

        const data = await response.json();
        return data.choices[0].message.content;
    }
}

// Main GIKIbot class
export class GIKIbot {
    constructor() {
        this.qaChain = null;
        this.vectorstore = null;
        this.processor = new GIKIDocumentProcessor();
        this.qualityChecker = new AnswerQualityChecker();
        this.initialized = false;
    }

    async initializeSystem() {
        try {
            if (await fs.pathExists(INDEX_PATH)) {
                // Load existing index (simplified - in real implementation you'd use a proper vector store)
                this.vectorstore = { loaded: true };
            } else {
                const documents = await this.processor.loadDocuments();
                if (documents.length === 0) {
                    return '❌ No documents found. Add files to the \'data\' folder.';
                }

                // Create vector store (simplified)
                this.vectorstore = { documents, loaded: true };
                await this.saveIndex();
            }

            this.initialized = true;
            return '✅ System ready! Ask questions now.';
        } catch (error) {
            return `❌ Error initializing system: ${error.message}`;
        }
    }

    async saveIndex() {
        // Simplified index saving
        await fs.ensureDir(INDEX_PATH);
        await fs.writeJson(path.join(INDEX_PATH, 'documents.json'), this.vectorstore.documents);
    }

    async loadIndex() {
        try {
            const documentsPath = path.join(INDEX_PATH, 'documents.json');
            if (await fs.pathExists(documentsPath)) {
                this.vectorstore = {
                    documents: await fs.readJson(documentsPath),
                    loaded: true
                };
                return true;
            }
        } catch (error) {
            console.error('Error loading index:', error.message);
        }
        return false;
    }

    async askQuestion(question) {
        if (!this.initialized) {
            return '⚠️ System not initialized yet.';
        }

        if (!question.trim()) {
            return '⚠️ Please enter a valid question.';
        }

        try {
            // Get answer from documents
            const answer = await this.getDocumentAnswer(question);
            
            // Assess answer quality
            console.log('🔍 Assessing answer quality...');
            const qualityAssessment = await this.qualityChecker.assessAnswerQuality(question, answer);
            
            console.log(`🔍 Overall score: ${qualityAssessment.overallScore.toFixed(2)}/10`);
            console.log(`🔍 AI assessment: ${qualityAssessment.aiAssessment.aiSufficient}`);
            console.log(`🔍 AI reason: ${qualityAssessment.aiAssessment.aiReason}`);

            const needsFallback = !qualityAssessment.isSufficient;

            if (needsFallback) {
                console.log('🔍 Fallback triggered!');
                const redditAnswer = await this.getRedditAnswer(question);
                if (redditAnswer) {
                    return `⚠️ Not found in official documents. Based on Reddit discussions:\n\n${redditAnswer}`;
                } else {
                    return `⚠️ ${answer}\n\n(No additional information found on Reddit)`;
                }
            } else {
                return answer;
            }

        } catch (error) {
            return `❌ Error: ${error.message}`;
        }
    }

    async getDocumentAnswer(question) {
        // Simplified document search and answer generation
        const documents = this.vectorstore.documents || [];
        
        // Simple keyword matching (in real implementation, use proper vector search)
        const relevantDocs = documents.filter(doc => 
            question.toLowerCase().split(' ').some(word => 
                doc.content.toLowerCase().includes(word)
            )
        ).slice(0, 5);

        if (relevantDocs.length === 0) {
            return 'I don\'t have that information in the provided documents.';
        }

        // Generate answer using OpenRouter API
        const context = relevantDocs.map(doc => doc.content).join('\n\n');
        const prompt = `You are a helpful assistant for GIKI (Ghulam Ishaq Khan Institute of Engineering Sciences and Technology).
Answer questions based on official GIKI documents: prospectus, fee structure, academic rules, and handbook.

Instructions:
- Answer based only on context
- If answer not found, say "I don't have that information in the provided documents"
- Be specific and cite document source when possible
- Maintain professional, student-friendly tone

Context:
${context}

Question: ${question}

Answer:`;

        const answer = await this.callOpenRouterAPI(prompt);
        
        // Add sources
        const sources = [...new Set(relevantDocs.map(doc => `📄 ${doc.metadata.source}`))];
        const sourceText = sources.length > 0 ? `\n\nSources:\n${sources.join('\n')}` : '';
        
        return `${answer}${sourceText}`;
    }

    async getRedditAnswer(question) {
        try {
            // Simplified Reddit search (in real implementation, use proper Reddit API)
            const searchTerms = question.toLowerCase().split(' ').slice(0, 3);
            
            // Mock Reddit posts for demonstration
            const mockPosts = [
                {
                    title: 'GIKI Hostel Information',
                    selftext: 'The boys hostel has good facilities and the food is decent. Rooms are shared but comfortable.',
                    url: 'https://reddit.com/r/giki/comments/example1'
                },
                {
                    title: 'Academic Calendar 2024',
                    selftext: 'The academic calendar for 2024 has been updated. Check the official website for details.',
                    url: 'https://reddit.com/r/giki/comments/example2'
                }
            ];

            if (mockPosts.length === 0) {
                return null;
            }

            const redditContext = mockPosts.map(post => 
                `**${post.title}**\n${post.selftext.substring(0, 500)}...\n(Source: ${post.url})`
            ).join('\n\n');

            const prompt = `Answer the following question using the Reddit discussions:\n\nQuestion: ${question}\n\nReddit Posts:\n${redditContext}\n\nAnswer:`;
            
            const redditAnswer = await this.callOpenRouterAPI(prompt);
            
            const redditSources = mockPosts.map(post => 
                `🌐 r/giki: ${post.title.substring(0, 60)}...`
            );
            
            return `${redditAnswer}\n\nReddit Sources:\n${redditSources.join('\n')}`;

        } catch (error) {
            console.error('Error getting Reddit answer:', error.message);
            return null;
        }
    }

    async callOpenRouterAPI(prompt) {
        const response = await fetch('https://openrouter.ai/api/v1/chat/completions', {
            method: 'POST',
            headers: {
                'Authorization': `Bearer ${OPENROUTER_API_KEY}`,
                'Content-Type': 'application/json',
                'HTTP-Referer': 'https://huggingface.co',
                'X-Title': 'GIKI-RAG-bot'
            },
            body: JSON.stringify({
                model: 'deepseek/deepseek-r1-0528-qwen3-8b:free',
                messages: [
                    {
                        role: 'system',
                        content: 'You are a helpful assistant for GIKI (Ghulam Ishaq Khan Institute of Engineering Sciences and Technology).'
                    },
                    {
                        role: 'user',
                        content: prompt
                    }
                ],
                temperature: 0.1
            })
        });

        if (!response.ok) {
            throw new Error(`OpenRouter API error: ${response.status}`);
        }

        const data = await response.json();
        return data.choices[0].message.content;
    }

    async rebuildIndex() {
        try {
            if (await fs.pathExists(INDEX_PATH)) {
                await fs.remove(INDEX_PATH);
            }
            return await this.initializeSystem();
        } catch (error) {
            return `Error rebuilding index: ${error.message}`;
        }
    }

    isInitialized() {
        return this.initialized;
    }
}

File: Giki Chatbot Redit/chatbot.py
import os
import pdfplumber
import docx
import re
import json
from pathlib import Path
from typing import List
import numpy as np
import praw

from langchain.docstore.document import Document
from langchain_community.vectorstores import FAISS
from langchain_community.embeddings import HuggingFaceEmbeddings
from langchain.text_splitter import RecursiveCharacterTextSplitter
from langchain.chains import RetrievalQA
from langchain.prompts import PromptTemplate
from langchain_openai import ChatOpenAI

# keyword extraction
from keybert import KeyBERT

# Import our intelligent quality checker
from answer_quality_checker import AnswerQualityChecker

INDEX_PATH = "faiss_index"

# -----------------------------
# OpenRouter API Setup
# -----------------------------
os.environ["OPENROUTER_API_KEY"] = os.getenv("OPENROUTER_API_KEY")
os.environ["OPENAI_API_KEY"] = os.environ["OPENROUTER_API_KEY"]
os.environ["OPENAI_API_BASE"] = "https://openrouter.ai/api/v1"
os.environ["OPENAI_API_HEADERS"] = '{"HTTP-Referer":"https://huggingface.co", "X-Title":"GIKI-RAG-bot"}'

# -----------------------------
# Reddit Setup (OAuth)
# -----------------------------
reddit = praw.Reddit(
    client_id=os.getenv("REDDIT_CLIENT_ID"),
    client_secret=os.getenv("REDDIT_CLIENT_SECRET"),
    refresh_token=os.getenv("REDDIT_REFRESH_TOKEN"),
    user_agent=os.getenv("REDDIT_USER_AGENT")
)

# -----------------------------
# Keyword Extractor
# -----------------------------
kw_model = KeyBERT("all-MiniLM-L6-v2")

def extract_keywords(query: str, top_k=3):
    keywords = kw_model.extract_keywords(query, keyphrase_ngram_range=(1,2), stop_words='english', top_n=top_k)
    return [kw for kw, score in keywords]


def search_reddit_semantic(query: str, subreddit="giki", top_n=5, embeddings_model=None):
    """
    Fetch latest posts from r/giki and return top_n most semantically similar ones.
    Uses keyword extraction for better search.
    """
    # Step 1: Extract clean keywords/phrases
    keyphrases = extract_keywords(query, top_k=3)
    search_terms = keyphrases if keyphrases else [query]

    print(keyphrases)


    posts = []
    try:
        for term in search_terms:
            for submission in reddit.subreddit(subreddit).search(term, limit=20):
                posts.append({
                    "title": submission.title,
                    "selftext": submission.selftext,
                    "url": submission.url
                })
    except Exception:
        return []

    if not posts or embeddings_model is None:
        return []

    # Combine title + body
    post_texts = [f"{p['title']}\n{p['selftext']}" for p in posts]

    # Compute embeddings
    post_embeddings = embeddings_model.embed_documents(post_texts)
    query_embedding = embeddings_model.embed_query(query)

    # Cosine similarity
    similarities = [
        np.dot(query_embedding, p_emb) / (np.linalg.norm(query_embedding) * np.linalg.norm(p_emb))
        for p_emb in post_embeddings
    ]

    # Top N results
    top_indices = np.argsort(similarities)[::-1][:top_n]
    return [posts[i] for i in top_indices]


# -----------------------------
# Document Processor
# -----------------------------
class GIKIDocumentProcessor:
    def __init__(self, data_folder="data"):
        self.data_folder = data_folder
        self.text_splitter = RecursiveCharacterTextSplitter(
            chunk_size=500,
            chunk_overlap=50,
            length_function=len,
            separators=["\n\n", "\n", ".", "!", "?", ",", " ", ""]
        )

    def extract_text_from_pdf(self, file_path: str) -> str:
        text = ""
        try:
            with pdfplumber.open(file_path) as pdf:
                for page_num, page in enumerate(pdf.pages):
                    page_text = page.extract_text()
                    if page_text:
                        page_text = re.sub(r'\s+', ' ', page_text)
                        text += f"\n[Page {page_num + 1}]\n{page_text}\n"
        except Exception:
            pass
        return text

    def extract_text_from_docx(self, file_path: str) -> str:
        text = ""
        try:
            doc = docx.Document(file_path)
            for paragraph in doc.paragraphs:
                if paragraph.text.strip():
                    text += paragraph.text + "\n"
        except Exception:
            pass
        return text

    def extract_text_from_txt(self, file_path: str) -> str:
        try:
            with open(file_path, 'r', encoding='utf-8') as file:
                return file.read()
        except Exception:
            return ""

    def extract_text_from_json(self, file_path: str) -> str:
        """Extract text from JSON files containing Reddit posts"""
        try:
            with open(file_path, 'r', encoding='utf-8') as file:
                data = json.load(file)
            
            text = ""
            if isinstance(data, list):
                for i, post in enumerate(data):
                    if isinstance(post, dict):
                        # Extract title and selftext
                        title = post.get('title', '')
                        selftext = post.get('selftext', '')
                        author = post.get('author', '')
                        post_id = post.get('id', '')
                        
                        # Combine post content
                        post_content = f"Title: {title}\nContent: {selftext}"
                        
                        # Add comments if available
                        comments = post.get('comments', [])
                        if comments:
                            post_content += "\nComments:\n"
                            for comment in comments:
                                if isinstance(comment, dict):
                                    comment_body = comment.get('body', '')
                                    comment_author = comment.get('author', '')
                                    if comment_body:
                                        post_content += f"- {comment_author}: {comment_body}\n"
                        
                        text += f"\n[Reddit Post {i+1} - ID: {post_id}]\n{post_content}\n"
            
            return text
        except Exception as e:
            # print(f"Error processing JSON file {file_path}: {str(e)}")
            return ""

    def load_documents(self) -> List[Document]:
        documents = []
        data_path = Path(self.data_folder)

        if not data_path.exists():
            return documents

        supported_extensions = {'.pdf', '.docx', '.txt', '.json'}

        for file_path in data_path.iterdir():
            if file_path.is_file() and file_path.suffix.lower() in supported_extensions:
                if file_path.suffix.lower() == '.pdf':
                    text = self.extract_text_from_pdf(str(file_path))
                elif file_path.suffix.lower() == '.docx':
                    text = self.extract_text_from_docx(str(file_path))
                elif file_path.suffix.lower() == '.txt':
                    text = self.extract_text_from_txt(str(file_path))
                elif file_path.suffix.lower() == '.json':
                    text = self.extract_text_from_json(str(file_path))
                else:
                    continue

                if text.strip():
                    chunks = self.text_splitter.split_text(text)

                    for i, chunk in enumerate(chunks):
                        if chunk.strip():
                            documents.append(
                                Document(
                                    page_content=chunk,
                                    metadata={
                                        "source": file_path.name,
                                        "chunk_id": i,
                                        "file_type": file_path.suffix.lower()
                                    }
                                )
                            )
        return documents


# -----------------------------
# bot
# -----------------------------
class GIKIbot:
    def __init__(self):
        self.qa_chain = None
        self.vectorstore = None
        self.processor = GIKIDocumentProcessor()
        self.quality_checker = AnswerQualityChecker()

        self.embeddings = HuggingFaceEmbeddings(
            model_name="sentence-transformers/all-MiniLM-L6-v2",
            model_kwargs={'device': 'cpu'},
            encode_kwargs={'normalize_embeddings': True}
        )

        self.prompt_template = """You are a helpful assistant for GIKI (Ghulam Ishaq Khan Institute of Engineering Sciences and Technology).
Answer questions based on official GIKI documents: prospectus, fee structure, academic rules, and handbook.

Instructions:
- Answer based only on context
- If answer not found, say "I don't have that information in the provided documents"
- Be specific and cite document source when possible
- Maintain professional, student-friendly tone

Context:
{context}

Question: {question}

Answer:"""

        self.custom_prompt = PromptTemplate(
            template=self.prompt_template,
            input_variables=["context", "question"]
        )

    def initialize_system(self):
        try:
            if os.path.exists(INDEX_PATH):
                self.vectorstore = FAISS.load_local(
                    INDEX_PATH,
                    self.embeddings,
                    allow_dangerous_deserialization=True
                )
            else:
                documents = self.processor.load_documents()
                if not documents:
                    return "❌ No documents found. Add files to the 'data' folder."

                self.vectorstore = FAISS.from_documents(documents, self.embeddings)
                self.vectorstore.save_local(INDEX_PATH)

            # Initialize LLM
            llm = ChatOpenAI(
                model="deepseek/deepseek-r1-0528-qwen3-8b:free",
                base_url="https://openrouter.ai/api/v1",
                api_key=os.getenv("OPENAI_API_KEY"),
                default_headers={
                    "HTTP-Referer": "https://huggingface.co",
                    "X-Title": "GIKI-RAG-bot"
                },
                temperature=0.1
            )

            self.qa_chain = RetrievalQA.from_chain_type(
                llm=llm,
                chain_type="stuff",
                retriever=self.vectorstore.as_retriever(
                    search_type="similarity",
                    search_kwargs={"k": 5}
                ),
                return_source_documents=True,
                chain_type_kwargs={"prompt": self.custom_prompt}
            )

            return "✅ System ready! Ask questions now."
        except Exception as e:
            return f"❌ Error initializing system: {str(e)}"
    def ask_question(self, question: str) -> str:
        if not self.qa_chain:
            return "⚠️ System not initialized yet."

        if not question.strip():
            return "⚠️ Please enter a valid question."

        try:
            response = self.qa_chain.invoke({"query": question})
            answer = response["result"]
            source_docs = response["source_documents"]

            # Use intelligent quality checker to assess answer
            #print("🔍 Assessing answer quality...")
            quality_assessment = self.quality_checker.assess_answer_quality(question, answer)
            
            # Debug: Print quality assessment details
          #  print(f"🔍 Overall score: {quality_assessment['overall_score']:.2f}/10")
          #  print(f"🔍 AI assessment: {quality_assessment['ai_assessment']['ai_sufficient']}")
          #  print(f"🔍 AI reason: {quality_assessment['ai_assessment']['ai_reason']}")
          #  print(f"🔍 Basic scores: {quality_assessment['basic_scores']}")
           
            needs_fallback = not quality_assessment['is_sufficient']
            

           # print(f"🔍 Fallback triggered! Overall score: {quality_assessment['overall_score']:.2f}")
          # print(f"🔍 Reason: {quality_assessment['ai_assessment']['ai_reason']}")

            if needs_fallback:
                # print("📡 Document answer insufficient, searching Reddit...")
                top_posts = search_reddit_semantic(question, embeddings_model=self.embeddings)
                if top_posts:
                    # Build Reddit context
                    reddit_context = "\n\n".join([
                        f"**{p['title']}**\n{p['selftext'][:500]}...\n(Source: {p['url']})"
                        for p in top_posts if p['selftext']
                    ])
                    
                    # Get Reddit-based answer
                    llm = ChatOpenAI(
                        model="deepseek/deepseek-r1-0528-qwen3-8b:free",
                        base_url="https://openrouter.ai/api/v1",
                        api_key=os.getenv("OPENAI_API_KEY"),
                        temperature=0.2
                    )
                    reddit_answer = llm.invoke(
                        f"Answer the following question using the Reddit discussions:\n\n"
                        f"Question: {question}\n\nReddit Posts:\n{reddit_context}\n\nAnswer:"
                    ).content
                    
                    # Build Reddit sources
                    reddit_sources = set()
                    for post in top_posts:
                        reddit_sources.add(f"🌐 r/giki: {post['title'][:60]}...")
                    
                    reddit_source_text = "\n\nReddit Sources:\n" + "\n".join(reddit_sources) if reddit_sources else ""
                    
                    return f"⚠️ Not found in official documents. Based on Reddit discussions:\n\n{reddit_answer}{reddit_source_text}"
                else:
                    # No Reddit posts found either - return original answer with document sources
                    sources = set()
                    for doc in source_docs:
                        sources.add(f"📄 {doc.metadata['source']}")
                    source_text = "\n\nSources:\n" + "\n".join(sources) if sources else ""
                    return f"⚠️ {answer}\n\n(No additional information found on Reddit){source_text}"
            else:
                # Good document answer - return with document sources
                sources = set()
                for doc in source_docs:
                    sources.add(f"📄 {doc.metadata['source']}")
                source_text = "\n\nSources:\n" + "\n".join(sources) if sources else ""
                return f"{answer}{source_text}"
            
        except Exception as e:
            return f"❌ Error: {str(e)}"

